applications/FluidDynamicsApplication/custom_elements/two_fluid_navier_stokes.cpp

```diff
@@ -1879,15 +1926,240 @@ void TwoFluidNavierStokes<TElementData>::ComputeSplitting(
+    GeometryType::ShapeFunctionsGradientsType EnrichedInterfaceShapeDerivativesNeg = rInterfaceShapeDerivatives;
+
+    for (unsigned int i = 0; i < rInterfaceShapeDerivatives.size(); ++i){
+        EnrichedInterfaceShapeDerivativesPos[i] = prod(enr_pos_interp, rInterfaceShapeDerivatives[i]);
+    }
+
+    for (unsigned int i = 0; i < rInterfaceShapeDerivatives.size(); ++i){
+        EnrichedInterfaceShapeDerivativesNeg[i] = prod(enr_neg_interp, rInterfaceShapeDerivatives[i]);
+    }
+
+    double positive_volume = 0.0;
+    double negative_volume = 0.0;
+    for (unsigned int igauss_pos = 0; igauss_pos < rData.w_gauss_pos_side.size(); ++igauss_pos){
+        positive_volume += rData.w_gauss_pos_side[igauss_pos];
+    }
+
+    for (unsigned int igauss_neg = 0; igauss_neg < rData.w_gauss_neg_side.size(); ++igauss_neg){
+        negative_volume += rData.w_gauss_neg_side[igauss_neg];
+    }
+    const double element_volume = positive_volume + negative_volume;
+
+    auto geom = this->GetGeometry();
+    const double h_elem = ElementSizeCalculator<Dim,NumNodes>::AverageElementSize(geom);
+
+    double cut_area = 0.0;
+    for (unsigned int gp = 0; gp < rInterfaceWeights.size(); ++gp){
+        cut_area += rInterfaceWeights[gp];
+    }
+
+    const double density = 1.0/(1.0/positive_density + 1.0/negative_density);
+    const double viscosity = 1.0/(1.0/positive_viscosity + 1.0/negative_viscosity);
+
+    // Stabilization parameters
+    const double cut_stabilization_coefficient = 1.0;
+    const double stab_c1 = 4.0;
+    const double stab_c2 = 2.0;
+    const double dyn_tau = rData.DynamicTau;
+
+    const double dt = rData.DeltaTime;
+
+    const auto v_convection = rData.Velocity - rData.MeshVelocity;
+
+    for (unsigned int gp = 0; gp < rInterfaceWeights.size(); ++gp){
+
+        Vector vconv = ZeroVector(Dim);
+        double positive_weight = 0.0;
+        double negative_weight = 0.0;
+
+        for (unsigned int j = 0; j < NumNodes; ++j){
+            for (unsigned int dim = 0; dim < Dim; ++dim){
+                vconv[dim] += (rEnrInterfaceShapeFunctionNeg(gp, j) + rEnrInterfaceShapeFunctionPos(gp, j))
+                    *v_convection(j,dim);
+            }
+            positive_weight += rEnrInterfaceShapeFunctionNeg(gp, j);
+            negative_weight += rEnrInterfaceShapeFunctionPos(gp, j);
+        }
+
+        const double v_conv_norm = norm_2(vconv);
+
+        const double penalty_coefficient = cut_stabilization_coefficient *
+            density * 1.0 / (dyn_tau * density / dt + stab_c1 * viscosity / h_elem / h_elem +
+                                stab_c2 * density * v_conv_norm / h_elem) * element_volume / cut_area;
+
+        for (unsigned int i = 0; i < NumNodes; ++i){
+
+            for (unsigned int j = 0; j < NumNodes; ++j){
+
+                const array_1d<double, 3> pressure_gradient_j = geom[j].GetValue(PRESSURE_GRADIENT);
```

Comment:
```suggestion
                const auto& r_pressure_gradient_j = geom[j].GetValue(PRESSURE_GRADIENT);
```
Avoid copying values.

Reply:
Done.